### PR TITLE
Add support for [[wiki-style]] links in markdown

### PR DESCRIPTION
**Pull request by iamrecursion**
1.  Support for [[wiki-style]] links ensures that they do not get broken
    by the prettier formatting by being wrapped over multiple lines.
2.  The contents of the link (between the `[[]]` brackets) are treated
    as raw text. This is because the various tools that rely on the
    `[[]]` syntax do not agree on what is allowable between them.

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

Closes #9275.

### CODE REVIEW

**Review by fisker — APPROVED** [review]
LGTM

**Comment by iamrecursion** [comment]
Thanks so much! Is there anything that needs doing before merge?

**Comment by fisker** [comment]
Maybe add more tests like

```markdown
[[a
b]]
```

```markdown
[[[a]]]
```

```markdown
[[[a]]
```

```markdown
[[a]]]
```

```markdown
[[a[b](http://www.example.com/)]]
```

**Comment by iamrecursion** [comment]
Done!

**Comment by iamrecursion** [comment]
Any ideas as to the failure on Node 14? I can't reproduce it locally as I'm on macOS, but I can't see any rhyme or reason for it succeeding on macOS and Windows but not on Linux.

**Comment by fisker** [comment]
`FULL_TEST=1 yarn test`

**Comment by fisker** [comment]
The test failed because

```markdown
[[a
b]]
```

ast changed.

**Comment by iamrecursion** [comment]
But isn't that expected in this case?

**Comment by fisker** [comment]
@iamrecursion  

If I understand this correctly, 

```js
[[a
b]]
```

is not a `wikiLink`, right?

**Comment by iamrecursion** [comment]
Ah, I see what you mean. In all honesty I'm not sure _what_ the most sensible behaviour is here, so following whatever happens with a standard `[]()` link is probably best. Can you point me to how this is handled for standard links?

**Comment by fisker** [comment]
I'm not sure if current output is expected, if it is, this is easy to fix, we just convert the first AST from `linkReference` in to `wikiLink`.

**Comment by fisker** [comment]
You can add something like

```js
  // `linkReference` with line breaks maybe covert to `wikiLink` after print
  if (ast.type === "linkReference" && ast.label.includes('\n')) {
    return {
		type: "wikiLink",
		value: ast.label
	      .trim()
	      .replace(/[\t\n ]+/g, " ")
	      .toLowerCase()
	};
  }
```
[here](https://github.com/prettier/prettier/blob/d096a34194f64e0bcf75e83769850de23e1b07c5/src/language-markdown/printer-markdown.js#L931) to pass test.

**Comment by fisker** [comment]
> If prettier doesn't turn the following into a `link`, then it shouldn't turn the test case into a `wikiLink`.
> 
> ```
> [foo bar
> baz](https://example.com)
> ```

This is not the failed case, [this one](https://github.com/prettier/prettier/pull/9275#issuecomment-699841614) is.

**Comment by iamrecursion** [comment]
> This is not the failed case, [this one](https://github.com/prettier/prettier/pull/9275#issuecomment-699841614) is.

I understand that, but my thinking is that the `wikiLink` test case should match the behaviour of of a standard link [like this](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBtAhgAgEYCcIDWCAOlADYCWUBAugBQBmEEO6uAlCADQgQAOMCtADOyUG3wB3AApsEolOjKT0AT1E886MERgBlPtqoBzZDFwBXODzgBbbHAAmjpwBl0UYxfTG4AMQhcW3QYQU9kEHQLGAhuEAALGFsyAHV4inhhQzA4PXkMigA3DNUIsGENECphOFwYaVwfYOQGJRqeACthAA8AIUadOH10WzhXKjgWtusQLu69EzI4AEULCHgpsnaQQ1wa3Ajg3AJHCEkoOL5cKhgUikcYeOQADgAGHiuIGpTGvgiruD7QqTHgARzW8Aa-AUkWEAFooHAnE44rg4OCKGiGk10JttjVbBQzJYZsJFisIZMkK0tjMYOhsHcHk8kAAmHjmdAUSieADCEFszRQgIArHELDUACoMhQ07aFKwASSgLlgejA1wEAEEVXoYKolni4ABfY1AA) where it doesn't get formatted.

**Comment by iamrecursion** [comment]
Okay, I've finally had a chance to look into this, and while the proposed band-aid fixes the aforementioned new test case, it causes breakage of `example-518.md` in the markdown tests. Instead I've settled for allowing the wikiLink regex to match over multiple lines, and stripping out the line breaks during parsing.

Please let me know what your thoughts are!

**Comment by iamrecursion** [comment]
Hey @fisker is there anything else I need to do to have this merged?

**Review by fisker — COMMENTED** [review]
LGTM

**Review by fisker — APPROVED** [review]
LGTM

**Comment by iamrecursion** [comment]
As I don't have write access you'll have to hit merge yourself I think @fisker!

File: src/language-markdown/parser-markdown.js
Comment (evilebottnawi) [inline src/language-markdown/parser-markdown.js:102]: Hope we don't break content using regexp here

**Review by thorn0 — CHANGES_REQUESTED** [review]
That's not what I expect from `--prose-wrap never`:

**Prettier pr-9275**
[Playground link](https://deploy-preview-9275--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAzAzgEwwAgNp4bYA6U6AhtgLpUgA0IEADjAJbTrKjkBOPEAdwAKvBJxTkANgPIBPTgwBGPcmADWcGAGUmq1lADmyGDwCucBnAC2iuJkx2AMuUOnyBuADEIPK+Rhshsgg5KYwEPQgABYwVpIA6lGs8Oi6YHBaYsmsAG7JssFg6Aog+uhwPDBCKgZ+yKhS5QwAVugAHgBCKuqaWuRWcI76cPWNFiCtbVr6BpJwAIqmEPCjkk0gujzlPMF+PGqYglCRTDz6MPGsmDBRyAAcAAwMpxDl8SpMwVBwORWRAI5LeDVZjiELoAC03zsdkiPDggNY8Oq7jqSAaa3G5SsrGMZixMzmi2WI3RYwYMHIiku11uSAATBSVKxJDMAMIQKxokBwdAAVkipnKABUqeIMescuYAJJQBywLRgM4sACCcq0MFkc1W5QAvrqgA)
```sh
--parser markdown
--prose-wrap never
```

**Input:**
```markdown
fsdfs [[fsdf
sadf]]
```

**Output:**
```markdown
fsdfs [[fsdf sadf]]

```

File: src/language-markdown/parser-markdown.js
Comment (iamrecursion) [reply src/language-markdown/parser-markdown.js:102]: Can you clarify more what you're worried about?

**Comment by iamrecursion** [comment]
This is what I was a little confused by. I'll look into it - sorry for missing that!

**Comment by iamrecursion** [comment]
@thorn0 I think this should address your concerns!

**Review by thorn0 — APPROVED** [review]
LGTM

**Comment by iamrecursion** [comment]
Great! What's the policy for merging things?

**Comment by thorn0** [comment]
> What's the policy for merging things?

PRs like this one need 3 approvals from the maintainers. For simpler changes 2 is enough.

**Comment by iamrecursion** [comment]
That makes perfect sense! Thanks. ^^

**Review by evilebottnawi — APPROVED** [review]

**Comment by iamrecursion** [comment]
Thank you to all of you for guiding me through this! Contributing to prettier has been a really good experience.

**Pull request merged**
